File: src/main/java/com/main/stat/StatSet.java
package com.main.stat;

import com.main.material.Material;

import java.util.Collection;
import java.util.HashMap;

/**
 * Created by Jaden on 9/17/2015.
 */
public class StatSet {
    private final HashMap<StatBase, Double> stats = new HashMap<>();

    public StatSet add(StatBase stat, double s) {
        this.stats.put(stat, s);
        return this;
    }
    public StatSet add(StatBase stat) {
        return this.add(stat, stat.getDefaultValue());
    }
    public StatSet copy(StatBase stat, StatSet other) {
        return this.add(stat, other.get(stat));
    }
    public double get(StatBase key) {
        return this.stats.getOrDefault(key, key.getDefaultValue());
    }
    public StatSet copy() {
        StatSet set = new StatSet();
        set.stats.putAll(this.stats);
        return set;
    }
    public StatSet scaled(int i) {
        StatSet ret = new StatSet();
        for(StatBase stat : this.getStatsRaw().keySet()) {
            double s = stat.scale(i, this.get(stat));
            ret.add(stat, s);
        }
        return ret;
    }
    public HashMap<StatBase, Double> getStatsRaw() {
        return this.stats;
    }
    public StatSet combine(Collection<StatSet> others) {
        StatSet ret = this.copy();
        for (StatBase stat : this.getStatsRaw().keySet()) {
            for (StatSet other : others) {
                ret.add(stat, stat.combine(ret.get(stat), other.get(stat)));
            }
        }
//        this.getStatsRaw().keySet().forEach(stat -> others.forEach(other -> ret.add(stat, stat.combine(ret.get(stat), other.get(stat)))));
        return ret;
    }
    public StatSet modified(Material resource) {
        StatSet ret = this.copy();
        this.getStatsRaw().keySet().forEach(stat -> ret.add(stat, stat.modify(resource, ret.get(stat))));
        return ret;
    }
}

File: src/main/kotlin/stat/StatSet.kt
package stat

import java.util.*

/**
 * Created by Jaden on 9/17/2015.
 */
class StatSet : Iterable<Map.Entry<StatBase<*>, Any>> {

    val statsRaw: MutableMap<StatBase<*>, Any> = HashMap()

    fun <T> add(statBase: StatBase<T>, value: T): StatSet {
        //TODO: Possibly check to see if that StatBase already exists in the map
        statsRaw.put(statBase, statBase.makeStat(value) as Any)
        return this
    }

    operator fun <T> get(statBase: StatBase<T>): T {
        @Suppress("UNCHECKED_CAST") //https://kotlinlang.org/docs/reference/inline-functions.html#reified-type-parameters
        return statsRaw[statBase] as T
    }

    operator fun <T> set(statBase: StatBase<T>, value: T): StatSet {
        return add(statBase, value)
    }

    fun <T> getModifiedValue(statBase: StatBase<T>): T {
        return statBase.getModifiedValue(this)
    }

    fun copy(): StatSet {
        val set = StatSet()
        set.statsRaw.putAll(this.statsRaw)
        return set
    }

    fun combine(otherSet: StatSet): StatSet {
        val retSet = StatSet()
        for ((statBase, value) in statsRaw) {
            @Suppress("UNCHECKED_CAST")
            retSet[statBase as StatBase<Any>] = statBase.combine(value, otherSet[statBase])
        }
        return retSet
    }

    override fun iterator(): Iterator<Map.Entry<StatBase<*>, Any>> = statsRaw.iterator()
}